### PR TITLE
Flush fix for SimpleLogProcessor

### PR DESCRIPTION
**Pull request by cijothomas**
SimpleLogProcessor's force_flush was incorrectly implemented as it did not do anything assuming there was nothing to flush. But the `SimpleLogProcessor` uses a queue with pub-sub mechanism to operate, so flush is not a no-op. This is handled correctly for `SimpleSpanProcessor`. 

Instead of borrowing the approach from `SimpleSpanProcessor`, I have refactored the `SimpleLogProcessor` to use no queue and instead use a simple Mutex protected exporter/shutdown. I feel this is sufficient (and want to port this to `SimpleSpanProcessor` as well), but would like to get feedback on the approach - was there some scenario which prompted the queue/pub-sub for SimpleProcessors? Or this approach is sufficient? 

There should not be any perf concerns as SimpleProcessors are used for learning/dev scenarios, and not for production. (Except when you are exporting to operating system native tracing like etw, user_events, but for them we have written RentrantProcessor separately.)

~Note: I temporarily added some extra dependencies to test this out. If the direction is okay, I'll do a cleanup and modify the tests to not rely on tracing and friends.~

### CODE REVIEW

**Comment by cijothomas** [comment]
If this approach is good enough, that can remove the "crossbeam-channel = { version = "0.5", optional = true }" dependency from sdk completely as well?

**Comment by lalitb** [comment]
Thanks for picking this up, and the InMemoryExporter tests :) So there are two options here

**Option 1**: Keep the existing queue based approach, and fix the [SimpleLogProcessor::force_flush](https://github.com/open-telemetry/opentelemetry-rust/blob/03b94b35c9f83389a101132c76c17a0020ed5c91/opentelemetry-sdk/src/logs/log_processor.rs#L87) to process the queue (as done for [spans](https://github.com/open-telemetry/opentelemetry-rust/blob/03b94b35c9f83389a101132c76c17a0020ed5c91/opentelemetry-sdk/src/trace/span_processor.rs#L174)). Which mean `tokio::log!()` ->  `Logger::emit()` -> `LogProcessor::emit()` will return immediately after writing to queue, and a separate background thread will process the queue. No locks, and lags in application thread. The queue is unbounded, which means it can continue to grow if the input throughput exceeds the processing capacity. But then, SimpleLogProcessor is not for production use.

**Option 2**: The approach in this PR to do a export from application thread, and so a mutex lock is required to ensure that `Exporter::export()` is not called concurrently from multiple application threads. 

While most of the language implementation (probably all) go with Option 2, I don't have strong opinion against Option 1 if we address the flushing issue.  Nothing wrong with the performance gain even for non-production usage :)

**Comment by cijothomas** [comment]
> Thanks for picking this up, and the InMemoryExporter tests :) So there are two options here
> 
> **Option 1**: Keep the existing queue based approach, and fix the [SimpleLogProcessor::force_flush](https://github.com/open-telemetry/opentelemetry-rust/blob/03b94b35c9f83389a101132c76c17a0020ed5c91/opentelemetry-sdk/src/logs/log_processor.rs#L87) to process the queue (as done for [spans](https://github.com/open-telemetry/opentelemetry-rust/blob/03b94b35c9f83389a101132c76c17a0020ed5c91/opentelemetry-sdk/src/trace/span_processor.rs#L174)). Which mean `tokio::log!()` -> `Logger::emit()` -> `LogProcessor::emit()` will return immediately after writing to queue, and a separate background thread will process the queue. No locks, and lags in application thread. The queue is unbounded, which means it can continue to grow if the input throughput exceeds the processing capacity. But then, SimpleLogProcessor is not for production use.
> 
> **Option 2**: The approach in this PR to do a export from application thread, and so a mutex lock is required to ensure that `Exporter::export()` is not called concurrently from multiple application threads. Simple approach, and without external dependency on `crossbeam-channel`.
> 
> While most of the language implementation (probably all) go with Option 2, I don't have strong opinion against Option 1 if we address the flushing issue. Nothing wrong with the performance gain even for non-production usage :)

I want to use Option 2 only, unless there is a strong reason not to.
1. It keeps the code a lot simpler, and reduces dependencies.
2. Is aligned with the spirit of SimpleProcessors original spec intent.
3. If someone really wants the user thread to return quickly, but does not want to pay for the buffer either, then they can use BatchProcessor with capacity=1. (Not a valid scenario in my opinion!)

**Comment by lalitb** [comment]
> If someone really wants the user thread to return quickly, but does not want to pay for the buffer either, then they can use BatchProcessor with capacity=1. (Not a valid scenario in my opinion!)

that's a good point. Good to get feedback on the reason behind original implementation :)

**Comment by TommyCpp** [comment]
Pub-sub based approach first introduced in #502 to address the performance issue of simple span processors. One thing worth noting is currently users who don't use async runtime in Rust cannot use `BatchSpanProcessor` so there were some argument to keep the `SimpleSpanProcessor` relatively performant.

**Review by shaun-cox — APPROVED** [review]

**Comment by cijothomas** [comment]
> Pub-sub based approach first introduced in #502 to address the performance issue of simple span processors. One thing worth noting is currently users who don't use async runtime in Rust cannot use `BatchSpanProcessor` so there were some argument to keep the `SimpleSpanProcessor` relatively performant.

Thanks for sharing the background! 
#502 does not seem to addressing any end user needs. What I understand is that, that PR addresses the "must not block" part from the spec, just for spec compliance purposes.
Also, the spec does not require SimpleSpanProcessor to not-block/high-perf. It only says the API/SDK should not block, and leaves out the SpanProcessors.

"_Some downstream SpanProcessors and subsequent SpanExporters called from this operation may be used for testing, proof-of-concept ideas, or debugging and may not be designed for production use themselves. They are not in the scope of this requirement and recommendation._"

@jtescher Can you comment on any actual use cases on having a performant simple span processor? Or we did that just for spec compliance?  If that is the case, would you be okay to revert that change, as spec does not require that from us.

**Comment by cijothomas** [comment]
> One thing worth noting is currently users who don't use async runtime in Rust cannot use BatchSpanProcessor so there were some argument to keep the SimpleSpanProcessor relatively performant.

This is something we need to address separately, anyways, do you agree? i.e We want to support users who do not use aync runtimes. (not just traces, but metrics too)

**Comment by hdost** [comment]
> Thanks for sharing the background!
#502 does not seem to addressing any end user needs. 

Not sure I would assume that.

**Comment by cijothomas** [comment]
> > Thanks for sharing the background!
> > #502 does not seem to addressing any end user needs.
> 
> Not sure I would assume that.

Can you share some user needs around SimpleProcessor? The linked PR talks mostly about spec compliance...

**Comment by jtescher** [comment]
Yeah I don't think it needs to be as performant 👍 . The simple case should be for scripts / jobs / other sync activities which shouldn't produce a volume of spans where it is impactful.

File: opentelemetry-sdk/src/logs/log_processor.rs [hunk 1]
```diff
@@ -42,62 +42,32 @@ pub trait LogProcessor: Send + Sync + Debug {
 /// emitted. If you find this limiting, consider the batch processor instead.
 #[derive(Debug)]
 pub struct SimpleLogProcessor {
-    sender: crossbeam_channel::Sender<Option<LogData>>,
-    shutdown: crossbeam_channel::Receiver<()>,
+    exporter: Mutex<Box<dyn LogExporter>>,
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply opentelemetry-sdk/src/logs/log_processor.rs:45]: Can I table this, and come back to this after this PR?

**Comment by cijothomas** [comment]
@TommyCpp @jtescher Please re-review. I have fixed handling the lock poisoning in latest commits.

The question from @shaun-cox about the very need of Mutex - Let me get back to it after this PR. 

This PR now makes it easy to write semi-integration tests for Logging using in-memory-exporter, and I want to add more of that.

**Review by jtescher — APPROVED** [review]

**Comment by cijothomas** [comment]
@hdost  Can you re-review as well as you have previously reviewed?

**Review by TommyCpp — APPROVED** [review]

**Comment by TommyCpp** [comment]
> @hdost Can you re-review as well as you have previously reviewed?

cc @hdost for final review

**Review by hdost — APPROVED** [review]

**Pull request merged**